### PR TITLE
8245245: WebSocket can lose the URL encoding of URI query parameters

### PR DESCRIPTION
**Pull request by Karm**
Proposes to backport [JDK-8245245](https://bugs.openjdk.org/browse/JDK-8245245).

The backport is clean as far as the actual `OpeningHandshake.java` goes. The test needed a little tweak so as to compile with `SimpleSSLContext` and also to handle the fact that the erroneous response does not bring a response body.

The test passes with the patch, fails without it.

```
$ make clean run-test TEST="jtreg:test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java"
...
==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR   
   jtreg:test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java
                                                         1     1     0     0   
==============================
TEST SUCCESS

Stopping sjavac server
Finished building targets 'clean run-test' in configuration 'linux-x86_64-normal-server-release'
```
In addition to that, I compiled and executed the original `WebSocketTest.java` reproducer found on  [JDK-8245245](https://bugs.openjdk.org/browse/JDK-8245245) JIRA.

## Unpatched Temurin-11.0.17+8  :x: 
```
$ java WebSocketTest 
Http Request
http://localhost:8000/?raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
Server RequestURI: /?raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
WebSocket Request
ws://localhost:8000/?&raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
Server RequestURI: /?&raw=abc+def/ghi=xyz&encoded=abc+def/ghi=xyz
```

## Patched jdk11u :heavy_check_mark: 
```
$ java WebSocketTest 
Http Request
http://localhost:8000/?raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
Server RequestURI: /?raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
WebSocket Request
ws://localhost:8000/?&raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
Server RequestURI: /?&raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
```
The patched version correctly leaves the latter part of the query param encoded.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8245245](https://bugs.openjdk.org/browse/JDK-8245245): WebSocket can lose the URL encoding of URI query parameters
 * [JDK-8298588](https://bugs.openjdk.org/browse/JDK-8298588): WebSockets: HandshakeUrlEncodingTest unnecessarily depends on a response body

### Reviewers
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**)
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk11u-dev pull/1558/head:pull/1558` \
`$ git checkout pull/1558`

Update a local copy of the PR: \
`$ git checkout pull/1558` \
`$ git pull https://git.openjdk.org/jdk11u-dev pull/1558/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1558`

View PR using the GUI difftool: \
`$ git pr show -t 1558`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk11u-dev/pull/1558.diff">https://git.openjdk.org/jdk11u-dev/pull/1558.diff</a>

</details>

### CODE REVIEW

**Comment by jerboaa** [comment]
@Karm Please change the PR title to `Backport c07ce7eec71aefbd3cb624e03ca53f5148d01f19` so that the bots recognize this as a backport. Please be sure to run `:jdk_net` test group before/after as well.

**Review by phohensee — CHANGES_REQUESTED** [review]
Looks good except for using getResponse().uri() and a few formatting issues in HandshakeUrlEncodingTest.java.

The change to use getResponse().uri() instead of getResponse().body() fixes a bug in tip, so can't be part of a backport. The thing to do is use the buggy code from the original commit in this backport, fix the problem in tip, then backport that fix to 19u, 17u, and 11u.

On formatting issues, future backports are easier if you match the original formatting as closely as possible. Looks like you're using and IDE that auto-formats.

At line 59:

import static java.lang.System.out;
import static java.net.http.HttpClient.Builder.NO_PROXY;
import static org.testng.Assert.*;

should be

import static java.net.http.HttpClient.Builder.NO_PROXY;
import static org.testng.Assert.*;
import static java.lang.System.out;

to more closely follow the original. In fact, if all the test uses is assertEquals, assertNotNull, and fail, just use the original commit code.

Indent/format the same as the original commit: lines 85-88, 89-92, 106-108, 111, 115, 133-135, 137, 141-146, 153-158, 174-176, 188-190, and missing blank line after 196. Might have missed somthing, but you get the idea.

**Comment by Karm** [comment]
@phohensee Thank you for the review.

Ad formatting: Ack. Will do.

Ad

> The change to use getResponse().uri() instead of getResponse().body() fixes a bug in tip, so can't be part of a backport.

Oh...I see. So the fact that the body is returned even though there was a failure is a bug, not a feature/desirable chage? I will browse the Jira and check the spec to make sure I am not bending the test to a faulty implementation then.


Cheers
Karm

**Comment by jerboaa** [comment]
> Oh...I see. So the fact that the body is returned even though there was a failure is a bug, not a feature/desirable chage? I will browse the Jira and check the spec to make sure I am not bending the test to a faulty implementation then.

Even if you discover a discrepancy, this would need to get fixed in [JDK head](https://github.com/openjdk/jdk) first and then backported. Please keep that in mind.

**Comment by Karm** [comment]
Helo, @phohensee,

I changed the behavior of the test in the tip (HEAD), so as I can then backport it along the JDK-8245245 patch to 19u, 17u, and 11u. 

https://github.com/openjdk/jdk/pull/11486

@jerboaa If I am reading the [spec](https://www.rfc-editor.org/rfc/rfc6455.html#section-5.5.1) right, there is IMHO no need to change the behavior of the implementation.

**Comment by Karm** [comment]
@jerboaa I realized the patch is already backported to JDK 19 and JDK 17. So it's just JDK 11 I'd like to backport JDK-8245245 to.   And IIUC, the change in https://github.com/openjdk/jdk/pull/11486 needed for JDK 11 will need backporting to JDK 19, JDK 17 too.

**Comment by jerboaa** [comment]
> @jerboaa I realized the patch is already backported to JDK 19 and JDK 17. So it's just JDK 11 I'd like to backport JDK-8245245 to. And IIUC, the change in [openjdk/jdk#11486](https://github.com/openjdk/jdk/pull/11486) needed for JDK 11 will need backporting to JDK 19, JDK 17 too.

Yes.

**Comment by jerboaa** [comment]
FWIW, since https://bugs.openjdk.org/browse/JDK-8240666 is not in 11u, it would be acceptable to alter the test to account for this. A comment in the test could explain that. I don't think backporting JDK-8240666 would be appropriate (behaviour change) and it looks like getting the test change (account for the optional body) in JDK head might not be wanted.

**Comment by dfuch** [comment]
It is important that the tests checks the URI received by the server, and that's probably why the body was used here. The server writes the URI it receives in the response body. This provides an end-to-end check that what was received is what we expected to send.

**Comment by Karm** [comment]
Hello @dfuch, @jerboaa, @phohensee, @jaikiran 

I'd like this tets change to go forward: https://github.com/openjdk/jdk/pull/11486
It correctly fails without JDK-8245245 and passes with it. It is not necessary to be checking the body.

When that test change is integrated, I'd initiate its backport.  Close this pr (https://github.com/openjdk/jdk11u-dev/pull/1558) and open a new one, that would just cleanly add JDK-8245245...


I hope that is all right process wise.

**Comment by dfuch** [comment]
> It correctly fails without JDK-8245245 and passes with it. It is not necessary to be checking the body.

Thanks for double checking. In that case I believed you're covered, I personally have no objection in that case. I mostly commented because I don't think there's a bug in the original test. It just uses a different method to check for success. As for approval I'll defer to the JDK 11 maintainers.

**Comment by Karm** [comment]
Hello @dfuch, @phohensee,

With jdk/pull/11486 merged, I would kindly ask for a review of this backport to JDK11u.
- the patch in OpeningHandshake.java did not require any changes
- the diff for the test between what is in mainline and what I propose here is this: https://editor.mergely.com/Hh1c3Z1N/

thx
K.

**Review by phohensee — APPROVED** [review]
Only a few format issues.

Lgtm with Severin's recommendations.

**Comment by Karm** [comment]
@phohensee @jerboaa Ready :+1:

**Comment by jerboaa** [comment]
@Karm For backport bugs please only `/integrate` once you've got approval for push to `jdk11u-dev` on the bug via the `jdk11u-fix-yes` label. See last point in: https://wiki.openjdk.org/display/JDKUpdates/How+to+contribute+or+backport+a+fix

If you don't yet have JBS access, please provide me with the relevant text and I'll add it on your behalf.

File: test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java
Comment (jerboaa) [inline test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java:127]: A better comment would be: `JDK-8240666 is not currently in 11u unlike later JDKs and we, therefore, need to check whether or not a body is present. This is OK as previous assertions verify a fix of JDK-8245245`. Something like that.

Comment (Karm) [reply test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java]: Ack.

**Comment by RealCLanger** [comment]
@Karm Thanks for contributing this backport to JDK 11 Updates. This looks good to me. As @jerboaa mentioned, the next step would be to add the fix request label and information to the JBS bug.

I'd have one further request: Can you please backport your fix for [JDK-8298588](https://bugs.openjdk.org/browse/JDK-8298588) to JDK 20 and JDK17 that we don't miss it out there. For JDK20, you can use the "/backport jdk20" comment on the commit in head. If the backport is clean, which I assume, you can right away integrate as it complies to the RDP1 rules (testfix). For jdk17u-dev, you'll need a PR and maintainer approval by labeling `jdk17u-fix-request`

**Comment by jerboaa** [comment]
> For JDK20, you can use the "/backport jdk20" comment on the commit in head.

I think that's only available for committers. Just FYI.

**Comment by Karm** [comment]
> > For JDK20, you can use the "/backport jdk20" comment on the commit in head.
> 
> I think that's only available for committers. Just FYI.

It is. I tried that the other day: https://github.com/openjdk/jdk/commit/c07ce7eec71aefbd3cb624e03ca53f5148d01f19#commitcomment-91811154

**Comment by RealCLanger** [comment]
OK, let me try to trigger the backports for you, see [here](https://github.com/openjdk/jdk/commit/909d0cb4d9475fd367b8bc64a6b50c5a324e9a01#commitcomment-93691813)...

**Comment by Karm** [comment]
Thanks, lemme process it today.

**Comment by Karm** [comment]
Hello, @jerboaa, This is the JBS text proposal, formatted for JIRA:

```

I would like [JDK-8245245|https://bugs.openjdk.org/browse/JDK-8245245] to get backported to JDK 11
so as more libraries can start using JDK's own WebSocket client instead of depending
on other implementations.

For instance, [Fabric8 Kubernetes client|https://github.com/fabric8io/kubernetes-client/blob/master/httpclient-jdk/README.md#jdk-client-for-fabric8]
has this very issue with JDK's WebSocket client and it uses OkHttp3 or Vert.x implementations instead.

I used these [JBang|https://www.jbang.dev/] scripts to briefly showcase that both Vert.x and OkHttp3 implementations
are fine running on JDK 11 and JDK 11's WebSocket client needs fixing: (It uses Undertow as the server)

[JDKClient.java|https://gist.github.com/Karm/58959f7bc1d3ef675eecd1e12e56094c]
[OKHttp3Client.java|https://gist.github.com/Karm/6ed845a22a8b8331f95292bb992ee7e0]
[VertXClient.java|https://gist.github.com/Karm/a04a778f352e8d29667eb38a219d4e4b]

{code}
$ java --version
openjdk 11.0.17 2022-10-18

$ ./JDKClient.java
[jbang] Building jar...
The query string was: &raw=abc+def/ghi=xyz&encoded=abc+def/ghi=xyz

$ ./OKHttp3Client.java
[jbang] Building jar...
The query string was: raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz

$ ./VertXClient.java
[jbang] Building jar...
The query string was: raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
{code}

Patched, see the JDK client fixed:
{code}
$ java --version
openjdk 11.0.255-internal 2023-01-17

$ ./JDKClient.java 
The query string was: &raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz

$ ./OKHttp3Client.java 
The query string was: raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz

$ ./VertXClient.java 
The query string was: raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz
{code}

Thanks
K.
```

**Review by jerboaa — APPROVED** [review]
Seems OK to me.

**Review by phohensee — APPROVED** [review]
Lgtm.

**Comment by GoeLin** [comment]
Hi @Karm,
I saw your JBS comment. Next time please reason why you consider the risk low.
Thanks.

**Comment by Karm** [comment]
> OK, let me try to trigger the backports for you, see [here](https://github.com/openjdk/jdk/commit/909d0cb4d9475fd367b8bc64a6b50c5a324e9a01#commitcomment-93691813)...

THX @RealCLanger 
JDK 17 https://github.com/openjdk/jdk17u-dev/pull/982
JDK 20 https://github.com/openjdk/jdk20/pull/76

**Comment by Karm** [comment]
> Hi @Karm, I saw your JBS comment. Next time please reason why you consider the risk low. Thanks.

@GoeLin Ack. Will do.

My own thinking is along the lines that it is a well isolated change in a very comprehensible part of code, it aligns with the spec and it does not bring any surprising behavior when compared to some popular WebSocket client libs used with JDK 11, as noted in https://github.com/openjdk/jdk11u-dev/pull/1558#issuecomment-1361109845.

I will try to be more eloquent about risks of particular backports in the it in future.

**Comment by RealCLanger** [comment]
> > OK, let me try to trigger the backports for you, see [here](https://github.com/openjdk/jdk/commit/909d0cb4d9475fd367b8bc64a6b50c5a324e9a01#commitcomment-93691813)...
> 
> THX @RealCLanger JDK 17 [openjdk/jdk17u-dev#982](https://github.com/openjdk/jdk17u-dev/pull/982) JDK 20 [openjdk/jdk20#76](https://github.com/openjdk/jdk20/pull/76)

Hi @Karm, I already did the backports, they are already integrated. You can close your PRs... 😄

**Comment by Karm** [comment]
> > > OK, let me try to trigger the backports for you, see [here](https://github.com/openjdk/jdk/commit/909d0cb4d9475fd367b8bc64a6b50c5a324e9a01#commitcomment-93691813)...
> > 
> > 
> > THX @RealCLanger JDK 17 [openjdk/jdk17u-dev#982](https://github.com/openjdk/jdk17u-dev/pull/982) JDK 20 [openjdk/jdk20#76](https://github.com/openjdk/jdk20/pull/76)
> 
> Hi @Karm, I already did the backports, they are already integrated. You can close your PRs... smile

@RealCLanger, Oh, I see :-D Sorry for the noise.

**Comment by phohensee** [comment]
@Karm, please add /integrate and I'll sponsor.

**Comment by Karm** [comment]
THX @phohensee  : )

**Pull request closed without merging**
